[Bug]: Non-default settings file values are not respected

### Has your issue already been fixed?

- [X] Have you checked to see if your issue still exists on the `master` branch? See [the docs](https://github.com/dosisod/refurb#developing) for instructions on how to setup a local build of Refurb.
- [X] Have you looked at the open/closed issues to see if anyone has already reported your issue?

### The Bug

Non-default values for `format`, `python_version`, or `sort_by` used in a settings file are not respected. Instead, they get overridden by the default values from parsing command line arguments. The documentation states:

> ...the values in the config file will be merged with the values specified via the command line.

...but even when there are no values specified for these specific options on the command line, the defaults are used anyway and stomp over the values provided in the config file. This can be seen by updating an existing test, `test_command_line_args_merge_config_file()`, to include the three affected options:

```python
def test_command_line_args_merge_config_file() -> None:
    contents = """\
[tool.refurb]
load = ["some", "folders"]
ignore = [100, "FURB101"]
enable = ["FURB111", "FURB222"]
quiet = true
format = "github"
python_version = "3.10"
sort_by = "error"
"""

    command_line_args = parse_args(
        ["--load", "x", "--ignore", "123", "--enable", "FURB200"]
    )
    config_file = parse_config_file(contents)

    merged = Settings.merge(config_file, command_line_args)

    assert merged == Settings(
        load=["some", "folders", "x"],
        ignore={ErrorCode(100), ErrorCode(101), ErrorCode(123)},
        enable={ErrorCode(111), ErrorCode(222), ErrorCode(200)},
        quiet=True,
        format="github",
        python_version=(3, 10),
        sort_by="error"
    )
```

Emits the following error (when run with a Python 3.11 interpreter and with very verbose `pytest` output):

```
___________________________________________________________ test_command_line_args_merge_config_file ___________________________________________________________

    def test_command_line_args_merge_config_file() -> None:
        contents = """\
    [tool.refurb]
    load = ["some", "folders"]
    ignore = [100, "FURB101"]
    enable = ["FURB111", "FURB222"]
    quiet = true
    format = "github"
    python_version = "3.10"
    sort_by = "error"
    """

        command_line_args = parse_args(
            ["--load", "x", "--ignore", "123", "--enable", "FURB200"]
        )
        config_file = parse_config_file(contents)

        merged = Settings.merge(config_file, command_line_args)

>       assert merged == Settings(
            load=["some", "folders", "x"],
            ignore={ErrorCode(100), ErrorCode(101), ErrorCode(123)},
            enable={ErrorCode(111), ErrorCode(222), ErrorCode(200)},
            quiet=True,
            format="github",
            python_version=(3, 10),
            sort_by="error"
        )
E       AssertionError: assert Settings(files=[], explain=None, ignore={ErrorCode(id=100, prefix='FURB', path=None), ErrorCode(id=123, prefix='FURB', path=None), ErrorCode(id=101, prefix='FURB', path=None)}, load=['some', 'folders', 'x'], enable={ErrorCode(id=111, prefix='FURB', path=None), ErrorCode(id=222, prefix='FURB', path=None), ErrorCode(id=200, prefix='FURB', path=None)}, disable=set(), debug=False, generate=False, help=False, version=False, quiet=True, enable_all=False, disable_all=False, config_file=None, python_version=(3, 11), mypy_args=[], format='text', sort_by='filename') == Settings(files=[], explain=None, ignore={ErrorCode(id=100, prefix='FURB', path=None), ErrorCode(id=123, prefix='FURB', path=None), ErrorCode(id=101, prefix='FURB', path=None)}, load=['some', 'folders', 'x'], enable={ErrorCode(id=111, prefix='FURB', path=None), ErrorCode(id=222, prefix='FURB', path=None), ErrorCode(id=200, prefix='FURB', path=None)}, disable=set(), debug=False, generate=False, help=False, version=False, quiet=True, enable_all=False, disable_all=False, config_file=None, python_version=(3, 10), mypy_args=[], format='github', sort_by='error')
E
E         Matching attributes:
E         ['files',
E          'explain',
E          'ignore',
E          'load',
E          'enable',
E          'disable',
E          'debug',
E          'generate',
E          'help',
E          'version',
E          'quiet',
E          'enable_all',
E          'disable_all',
E          'config_file',
E          'mypy_args']
E         Differing attributes:
E         ['python_version', 'format', 'sort_by']
E
E         Drill down into differing attribute python_version:
E           python_version: (3, 11) != (3, 10)
E           At index 1 diff: 11 != 10
E           Full diff:
E           - (3, 10)
E           ?      ^
E           + (3, 11)
E           ?      ^
E
E         Drill down into differing attribute format:
E           format: 'text' != 'github'
E           - github
E           + text
E
E         Drill down into differing attribute sort_by:
E           sort_by: 'filename' != 'error'
E           - error
E           + filename

test/test_arg_parsing.py:198: AssertionError
```

But it should not be emitting an error instance because values that are provided in a config file and not on the command line should be used.


### Version Info

```shell
Refurb: v1.15.0
Mypy: v1.2.0
```


### Python Version

Python 3.11.3

### Config File

```TOML
The bug manifests when any of these three lines are in the `pyproject.toml` file:


[tool.refurb]
format = "github"
python_version = "3.8"
sort_by = "error"
```


### Extra Info

Just curious...but why is argument parsing done manually instead of with a library, like `argparse`?

--- a/refurb/checks/builtin/use_bit_count.py
+++ b/refurb/checks/builtin/use_bit_count.py
@@ -45,7 +45,7 @@ class ErrorInfo(Error):
 
 
 def check(node: CallExpr, errors: list[Error], settings: Settings) -> None:
-    if settings.python_version < (3, 10):
+    if settings.get_python_version() < (3, 10):
         return  # pragma: no cover
 
     match node:

--- a/refurb/checks/builtin/use_isinstance_tuple.py
+++ b/refurb/checks/builtin/use_isinstance_tuple.py
@@ -54,7 +54,9 @@ def check(node: OpExpr, errors: list[Error], settings: Settings) -> None:
             and is_equivalent(lhs_args[0], rhs_args[0])
         ):
             type_args = (
-                "y | z" if settings.python_version >= (3, 10) else "(y, z)"
+                "y | z"
+                if settings.get_python_version() >= (3, 10)
+                else "(y, z)"
             )
 
             errors.append(

--- a/refurb/checks/datetime/simplify_fromisoformat.py
+++ b/refurb/checks/datetime/simplify_fromisoformat.py
@@ -66,7 +66,7 @@ def is_utc_timezone(timezone: str) -> bool:
 
 
 def check(node: CallExpr, errors: list[Error], settings: Settings) -> None:
-    if settings.python_version < (3, 11):
+    if settings.get_python_version() < (3, 11):
         return
 
     match node:

--- a/refurb/checks/functools/use_cache.py
+++ b/refurb/checks/functools/use_cache.py
@@ -40,7 +40,7 @@ class ErrorInfo(Error):
 
 
 def check(node: Decorator, errors: list[Error], settings: Settings) -> None:
-    if settings.python_version < (3, 9):
+    if settings.get_python_version() < (3, 9):
         return  # pragma: no cover
 
     match node:

--- a/refurb/main.py
+++ b/refurb/main.py
@@ -146,7 +146,7 @@ def run_refurb(settings: Settings) -> Sequence[Error | str]:
     opt.cache_fine_grained = True
     opt.allow_redefinition = True
     opt.local_partial_types = True
-    opt.python_version = settings.python_version
+    opt.python_version = settings.get_python_version()
 
     try:
         result = build(files, options=opt)

--- a/refurb/settings.py
+++ b/refurb/settings.py
@@ -37,10 +37,10 @@ class Settings:
     enable_all: bool = False
     disable_all: bool = False
     config_file: str | None = None
-    python_version: tuple[int, int] = get_python_version()
+    python_version: tuple[int, int] | None = None
     mypy_args: list[str] = field(default_factory=list)
-    format: Literal["text", "github"] = "text"
-    sort_by: Literal["filename", "error"] = "filename"
+    format: Literal["text", "github", None] | None = None
+    sort_by: Literal["filename", "error"] | None = None
 
     def __post_init__(self) -> None:
         if self.enable_all and self.disable_all:
@@ -77,11 +77,14 @@ class Settings:
             enable_all=old.enable_all or new.enable_all,
             quiet=old.quiet or new.quiet,
             config_file=old.config_file or new.config_file,
-            python_version=new.python_version,
+            python_version=new.python_version or old.python_version,
             mypy_args=new.mypy_args or old.mypy_args,
-            format=new.format,
-            sort_by=new.sort_by,
+            format=new.format or old.format,
+            sort_by=new.sort_by or old.sort_by,
         )
+
+    def get_python_version(self) -> tuple[int, int]:
+        return self.python_version or get_python_version()
 
 
 ERROR_ID_REGEX = re.compile("^([A-Z]{3,4})?(\\d{3})$")
@@ -158,10 +161,8 @@ T = TypeVar("T")
 def pop_type(  # type: ignore[misc]
     ty: type[T], type_name: str = ""
 ) -> Callable[..., T]:
-    def inner(  # type: ignore[misc]
-        config: dict[str, Any], name: str, *, default: T | None = None
-    ) -> T:
-        x = config.pop(name, default or ty())
+    def inner(config: dict[str, Any], name: str) -> T:  # type: ignore[misc]
+        x = config.pop(name, ty())
 
         if isinstance(x, ty):
             return x
@@ -208,18 +209,16 @@ def parse_config_file(contents: str) -> Settings:
     mypy_args = pop_list(config, "mypy_args")
     settings.mypy_args = [str(x) for x in mypy_args]
 
-    version = pop_str(config, "python_version")
-    settings.python_version = (
-        parse_python_version(version) if version else get_python_version()
-    )
+    if "python_version" in config:
+        version = pop_str(config, "python_version")
 
-    settings.format = validate_format(
-        pop_str(config, "format", default="text")
-    )
+        settings.python_version = parse_python_version(version)
 
-    settings.sort_by = validate_sort_by(
-        pop_str(config, "sort_by", default="filename")
-    )
+    if "format" in config:
+        settings.format = validate_format(pop_str(config, "format"))
+
+    if "sort_by" in config:
+        settings.sort_by = validate_sort_by(pop_str(config, "sort_by"))
 
     amendments: list[dict[str, Any]] = config.pop("amend", [])  # type: ignore
 

--- a/test/test_arg_parsing.py
+++ b/test/test_arg_parsing.py
@@ -183,6 +183,10 @@ load = ["some", "folders"]
 ignore = [100, "FURB101"]
 enable = ["FURB111", "FURB222"]
 quiet = true
+format = "github"
+sort_by = "error"
+python_version = "3.7"
+mypy_args = ["some", "args"]
 """
 
     command_line_args = parse_args(
@@ -197,6 +201,10 @@ quiet = true
         ignore={ErrorCode(100), ErrorCode(101), ErrorCode(123)},
         enable={ErrorCode(111), ErrorCode(222), ErrorCode(200)},
         quiet=True,
+        format="github",
+        sort_by="error",
+        python_version=(3, 7),
+        mypy_args=["some", "args"],
     )
 
 

